src/circuit/circuit_builder.rs

```diff
@@ -301,6 +349,107 @@ type SchedulerEventHandler = Box<dyn Fn(&SchedulerEvent)>;
+/// data by-value when possible while falling back to pass-by-reference otherwise.  In a
+/// synchronous circuit, each operator reads its input stream precisely once in each clock cycle.
+/// It is therefore possible to determine the last consumer at each clock cycle and give it the
+/// owned value from the channel.  It is furthermore possible for the scheduler to schedule
+/// operators that strongly prefer owned values last.
+///
+/// We capture ownership preferences at two levels.  First, each individual operator that consumes
+/// one or more streams exposes its preferences on a per-stream basis via an API method
+/// (e.g., `[UnaryOperator::input_preference]`).  The [`Circuit`] API allows the circuit builder
+/// to override these preferences when instantiating an operator, taking into account circuit topology
+/// and workload.  We express preference as a numeric value.
+///
+/// These preferences are associated with each edge in the circuit graph.  The schedulers we have
+/// built so far implement a limited form of ownership-aware scheduling.  They only consider strong
+/// preferences (`OwnershipPreference::require_owned` and stronger) and model them internally as hard
+/// constraints that must be satisfied for the circuit to be schedulable.  Weaker preferences are
+/// ignored.
+#[repr(transparent)]
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+pub struct OwnershipPreference(usize);
+
+impl OwnershipPreference {
+    /// Create a new instance with given numeric preference value (higher
+    /// value means stronger preference).
+    pub fn new(val: usize) -> Self {
+        Self(val)
+    }
+
+    /// The operator does not gain any speed up from consuming an owned value.
+    pub fn dont_care() -> Self {
+        Self::new(0)
+    }
+
+    /// The operator is likely to run faster provided an owned input.
+    ///
+    /// Preference levels above `prefer_owned` should not be used by operators and
+    /// are reserved for use by circuit builders through the `Circuit` API.
+    pub fn prefer_owned() -> Self {
+        Self::new(50)
+    }
+
+    /// The circuit will suffer a significant performance hit if the operator
+    /// cannot consume data in the channel by-value.
+    pub fn require_owned() -> Self {
+        Self::new(100)
+    }
+
+    /// Returns the numeric value of the preference.
+    pub fn numeric(&self) -> usize {
+        self.0
+    }
+}
+
+impl Display for OwnershipPreference {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if *self == Self::dont_care() {
+            f.write_str("?")
+        } else if *self == Self::prefer_owned() {
+            f.write_str("prefer_owned")
+        } else if *self == Self::require_owned() {
+            f.write_str("require_owned")
+        } else {
+            write!(f, "preference({})", self.numeric())
+        }
+    }
+}
+
+/// An edge in a circuit graph represents a stream connecting two
```

Comment:
"Edges represent scheduling dependencies. These are of two kinds: actual data dependencies through streams and scheduling dependencies induced by preference constraints."

Reply:
Let's call this Dependency, not Edge.

Reply:
I don't see what's wrong with the comment or the name `Edge`. An edge is a stream or a dependency. Your terminology makes sense too, but it's not obviously more correct.

Reply:
Edge is a much more abstract concept. These are edges that represent some specific relations in the circuit graph.